readme.md

```diff
@@ -44,7 +44,7 @@ Yields:
 
 ## API
 
-### visit([node](https://github.com/wooorm/unist#unist-nodes)\[, type\], callback\[, reverse\])
+### visit([node](https://github.com/wooorm/unist#unist-nodes)\[, type], callback\[, reverse])
```

Comment:
1) **Compilers.** What you are proposing effectively relies on undocumented (AFAIK) [compiler feature](https://github.com/wooorm/remark-html/blob/cd388f9b3878a4b608b4b4489a061fada558aa28/lib/compilers.js#L57-L62). It is not baked in since compilers are detached from `remark` core. Although it can be enforced by specifying some policy that compilers should follow, it seems wrong because it introduces irregularities to the design without much benefit.

(And BTW, `remark-man` doesn't seem to follow this convention.)

![](http://i.imgur.com/E3qiSXd.png)

2)  **Tool breakage.** Don't you think it's just wrong to create a `linkReference` node when the user means `text`? It may look as expected for human consumers, but utilities and plugins would be totally mislead by this and produce wrong output.

Among others, remark-lint's [no-undefined-references](https://github.com/wooorm/remark-lint/blob/master/lib/rules/no-undefined-references.js) and [no-unused-definitions](https://github.com/wooorm/remark-lint/blob/master/lib/rules/no-unused-definitions.js) would break since they won't be able to tell what is a shortcut reference and what is not.

IMO AST should not contain such ambiguities. Nice and clear AST interface is exactly what make `remark` great by enabling all kinds of plugins.

3) **Unsustainability.** Yeah, maybe I'm wrong with this, again, but I just wanted to expand a bit more.

I was talking about long documents evolving with time. With proposed feature, `remark` would remove escaping from undefined shortcut references (even if there was an escape mark originally), and this would lead to problems in the long run. Every time people introduce new definitions, they would have to check if there was an undefined reference before and escape it. (Which can be in itself not trivial, because of case-insensitivity.) And they can't take any definition away, even for one moment (because if they do, the next moment `remark` will ferociously remove any escaping from bracketed terms that happen to collide with that definition). Now add autosaving to the picture…

Anyway, this feature seems to be of a very non-local kind and this is what I don't really like about it.

4) **Usage.** Lastly, what is the use case? Square brackets are mostly used in technical texts (your example included), and `text` is just not an adequate type for this — of course you will need to escape!	`text` is designed for texts in natural language, not arbitrary mix of alphanumerals and punctuation:

```js
# foo.make\_seq(init\[, fn=x=>x\*x\*\~~x], \*args, \*\*kwargs)
```

Use `inlideCode` instead:


```js
# `foo.make_seq(init[, fn=x=>x*x*~~x], *args, **kwargs)`
```

In literature, square brackets are used very sparingly, perhaps most commonly to indicate text omissions. But this is a rare case (at least in normal texts), so I think it's not a problem to escape when needed.

Please correct me if I'm wrong or delusional…